### PR TITLE
[ORA-1931] Ensure deterministic order in GetAndUpdateActiveTopicWeights

### PR DESCRIPTION
**Pull request by guilherme-brandao**
## What is the purpose of the change

Fix potential non deterministic order in GetAndUpdateActiveTopicWeights function

## Testing and Verifying

*(Please pick one of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added unit test that validates ...*
  - *Added integration tests for end-to-end deployment with ...*
  - *Extended integration test for ...*
  - *Manually verified the change by ...*

## Documentation and Release Note

  - [ ] Does this pull request introduce a new feature or user-facing behavior changes?

Where is the change documented?
  - [ ] Specification (`x/{module}/README.md`)
  - [ ] Allora documentation site `docs.allora.network` source code at: `https://github.com/allora-network/docs`
  - [ ] Code comments?
  - [ ] N/A

### CODE REVIEW

**Review by kpeluso — DISMISSED** [review]
the true fix here is to:
1. store a map from block height to a list of active topics (also keep the current key store of active topics, which would still be used for queries)
2. when a topic is activated, append it to the list at the block height of its next epoch/nonce. if the list is full, then boot the topic of lowest weight (maybe precompute this minimum as we go)
3. within each endblock where its epoch ends, decide if the topic is still active and add it to the list of the next block height (nonce + EpochLength)
4. prune this new store at the current block -- by this point, only future blocks should be keys in this new data structure

We could also consider applying a cap to this list, and giving topics opportunity to boot out other topics via external tx or during EndBlock

File: x/emissions/module/rewards/rewards_internal.go [hunk 1]
```diff
@@ -450,7 +450,7 @@ func GetAllReputersOutput(
 	// finalScores := make([]alloraMath.Dec, numReputers)
 	newScores := make([]alloraMath.Dec, numReputers)
 
-	for maxGradient.Gt(maxGradientThreshold) && i < gradientDescentMaxIters {
+	for maxGradient.Gte(maxGradientThreshold) && i < gradientDescentMaxIters {
```
[the remark this answers is not in the document]
Comment (guilherme-brandao) [reply x/emissions/module/rewards/rewards_internal.go:453]: hmm I also added this to be in line with diederik's simulator

File: x/emissions/module/rewards/topic_rewards.go [hunk 2]
```diff
@@ -235,10 +239,10 @@ func GetAndUpdateActiveTopicWeights(
 			return errors.Wrapf(err, "failed to reset topic fee revenue")
 		}
 
-		// If the topic is inactive, add it to the list of inactive topics
+		// If the topic is inactive, inactivate it
 		if weight.Lt(moduleParams.MinTopicWeight) {
-			nowInactiveTopics = append(nowInactiveTopics, topic.Id)
-			return nil
+			err := k.InactivateTopic(ctx, topic.Id)
```
[the remark this answers is not in the document]
Comment (kpeluso) [reply x/emissions/module/rewards/topic_rewards.go:276]: not being edited from source as the iteration progresses here/anymore

**Review by relyt29 — APPROVED** [review]
Skimmed but looks good

**Pull request merged**